File: py/day07.py
import os

# https://adventofcode.com/2022/day/7 - Day 7: No Space Left On Device

test_input = """$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k"""

def add_dir(nm: str, parent):
    dir = { 'name': nm, 'parent': parent, 'dirs': {}, 'files': [] }
    if parent != None:
        parent['dirs'][nm] = dir
    return dir

def add_file(sz: int, name: str, parent):
    f = { 'name': name, 'size': sz }
    parent['files'].append(f)

def cd(to: str, current):
    if to == '/':
        dir = current
        while dir['parent'] != None:
            dir = dir['parent']
        return dir
    if to == '..':
        return current['parent']
    return current['dirs'][to]

def dir_size(dir, sizes, path):
    sz = 0
    dname = dir['name']
    fullname = path + '/' + dname

    for f in dir['files']:
        sz += f['size']
    for _,d in dir['dirs'].items():
        dsz = dir_size(d, sizes, fullname)
        sz += dsz
    sizes[fullname] = sz
    return sz

#lines = [line.strip().split() for line in test_input.splitlines()]
with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day07-input", "r") as f:
    lines = [line.strip().split() for line in f.readlines()]

# Build it
root = add_dir('root', None)
cwd = root
for line in lines[2:]:
    match line[0]:
        case '$':
            match line[1]:
                case 'cd':
                    cwd = cd(line[2], cwd)
                #case 'ls':
                #    print("ignore ls")
        case 'dir':
            # dir <dir-name>
            add_dir(line[1], cwd)
        case _:
            # file size and name
            add_file(int(line[0]), line[1], cwd)

# Analyse it
dir_sizes = {}
dir_size(root, dir_sizes, '')

free_space = 70000000 - dir_sizes['/root']
required = 30000000 - free_space

# print("Required extra space ", required)
# for path, sz in dir_sizes.items():
#     if sz >= needed:
#         print("Found ", path, sz)

print("Part 1: ", sum([sz for sz in dir_sizes.values() if sz <= 100000]))
print("Part 2: ", min([sz for sz in dir_sizes.values() if sz >= required]))

File: cs/days/day07.cs
namespace Shunty.AoC;

public class Day07 : AocDaySolution
{
    internal readonly record struct AoCFile(string Name, Int64 Size);

    internal class AoCDirectory
    {
        public string Name { get; init; }
        public AoCDirectory? Parent { get; init; }
        private List<AoCDirectory> _dirs = new();
        private List<AoCFile> _files = new();
        public IEnumerable<AoCDirectory> Directories => _dirs;
        public IEnumerable<AoCFile> Files => _files;

        public AoCDirectory(string name, AoCDirectory? parent)
        {
            Name = name;
            Parent = parent;
            Parent?._dirs.Add(this);
        }

        public AoCDirectory AddDirectory(string name) => new AoCDirectory(name, this);

        public void AddFile(string name, Int64 size) => _files.Add(new AoCFile(name, size));

        public AoCDirectory FindDirectory(string name)
        {
            return name == ".."
                ? this.Parent ?? this
                : _dirs.FirstOrDefault(d => d.Name == name) ?? this;
        }
    }

    private Int64 GetDirectorySizes(AoCDirectory dir, Dictionary<string, Int64> sizes, string path = "")
    {
        var fullpath = dir.Name == "/" ? "/" : $"{path}{dir.Name}/";
        var sz = dir.Files.Sum(f => f.Size);
        foreach (var d in dir.Directories)
        {
            sz += GetDirectorySizes(d, sizes, fullpath);
        }
        sizes[fullpath] = sz;
        return sz;
    }

    private const Int64 TotalDiskSpace = 70000000L;
    private const Int64 RequiredDiskSpace = 30000000L;

    public async Task Run()
    {
        var data = await AocUtils.GetDayLines(7);
        var root = BuildFileSystem(data);

        var dirsizes = new Dictionary<string, Int64>();
        GetDirectorySizes(root, dirsizes);
        var freespace = TotalDiskSpace - dirsizes["/"];
        var needed = RequiredDiskSpace - freespace;

        Console.WriteLine("Day 7");
        Console.WriteLine($"  Part 1: {dirsizes.Values.Where(v => v <= 100000).Sum()}");
        Console.WriteLine($"  Part 2: {dirsizes.Values.Where(v => v >= needed).Min()}");
    }

    private AoCDirectory BuildFileSystem(IEnumerable<string> input)
    {
        var root = new AoCDirectory("/", null);
        var cwd = root;
        foreach (var line in input.Skip(2))
        {
            var splits = line.Split(' ');
            switch (splits[0])
            {
                case "$":
                    if (splits[1] == "cd")
                    {
                        cwd = cwd.FindDirectory(splits[2]);
                    }
                    break;
                case "dir":
                    cwd.AddDirectory(splits[1]);
                    break;
                default:
                    cwd.AddFile(splits[1], Int64.Parse(splits[0]));
                    break;
            }
        }
        return root;
    }
}
